### ISSUE TITLE
Pairing worker with task priority?

### ISSUE THREAD

**Issue by jedie** [coleifer/huey#671]
Is it possible to setup some worker that should only consume tasks with a higher priority than X ?

The Problem is: If there are 4 workers and all of them are processing a long running tasks, than there is no worker for processing short, high priority tasks.

But if there is a way to split the worker into "process low priority task" and "process only high priority tasks" and if the task priority is choose by running time, then there is no blocking problematic...

**Comment by coleifer** [comment]
For those kinds of situations there may be some workarounds you can try, but nothing out-of-the-box that Huey could support.

Ideas:

* use a lock or locks. Downside: overly pessimistic / reduced throughput.
* implement your own semaphore that relies on your storage layer's atomic increment/decrement operations. Downside: write your own code.
* use a dedicated `Huey` instance for high-priority tasks. Downside: you'll need to run two consumers.

If you're using Redis, implementing a semaphore should be fairly easy, I might take a crack at it.

**Comment by coleifer** [comment]
I added a very simple lock implementation that can be acquired multiple times w/redis -- but it does not handle blocking/waiting or notifying. It's just a lock that can be acquired multiple times -- you'd want to use it in conjunction with Huey's built-in retry mechanism.

```python
from huey.contrib.helpers import lock_task_semaphore

huey = RedisHuey(...)

@huey.task(retries=2, retry_delay=60)
@lock_task_semaphore(huey, 'my-lock', value=3)  # Allow lock to be acquired up to 3 times.
def my_task(...):
    ...
```

Not sure if this helps address your issue, but you would use the `@lock_task_semaphore(huey, 'my-lock', value=3)` around the task or tasks that need to be limited to a certain number of concurrent executions.

**Comment by jedie** [comment]
For clarify:
There are two different task types:

1. periodic background tasks 
2. Task started by user

Some of these periodic background tasks (1.) takes ~1hour. The number of long-running tasks is greater than the number of workers. It's ok that they all together take more than one hour.

The Problem is: The user tasks (2.) should always be prioritized. They should not be blocked by the long running tasks.

> I added a very simple lock implementation that can be acquired multiple times w/redis

I think `@lock_task_semaphore()` can not help here.


> Ideas:
> 
>     * use a lock or locks. Downside: overly pessimistic / reduced throughput.
>     * implement your own semaphore that relies on your storage layer's atomic increment/decrement operations. Downside: write your own code.
>     * use a dedicated `Huey` instance for high-priority tasks. Downside: you'll need to run two consumers.

Think only the "run two consumers" can fix the issues. 
I use Huey with Django and start the consumer via "run_huey" manage command. Would be great if it's possible to "split" the worker via manage command arguments.

I take a look into Consumer() class and the doc here: https://huey.readthedocs.io/en/latest/consumer.html#options-for-the-consumer ... currently it's not possible to specify that a consumer should only run task with priority from X to Y...

But there is the "--no-periodic" switch... Idea: Start worker that only consume periodic tasks and worker that only consume non-periodic tasks... This will also fix the problem.

So i see two solutions:

1. Split worker by task priority
2. Split worker by periodic/non-periodic tasks

I am unsure which would be the best solution and how much effort is behind it.

**Comment by coleifer** [comment]
Thanks @pablop94 - I think this is what @jedie probably needs -- some way to use multiple queues with Django.
